In [1]:
import pandas as pd
import numpy as np
import os
from matplotlib import pyplot as plt
import numpy as np
import math
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA 



df = pd.read_csv('../capstone-projects/speed-dating-experiment/Speed Dating Data.csv')


In [2]:
#import seabord and matplotlib modules
import seaborn as sns
import matplotlib.pyplot as plt
#makes visualizations appear inline with code in notebook
% matplotlib inline

In [3]:
sub = df[["gender","imprelig", "goal", "date","go_out", 'exphappy', 'expnum','attr1_1', 'sinc1_1', 'intel1_1', 'fun1_1', 'amb1_1', 'shar1_1',"match","samerace"]]

In [4]:
match_diff_rc_mask = (sub["match"] == 1) & (sub["samerace"] == 0)

sub["match_diff_race"] = "NaN"



/Users/ltdanp21/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  app.launch_new_instance()


In [5]:
#create "new target" column based on values of two columns:

# create a mask that meets your condition
mask = ((sub["match"] == 1) & (sub["samerace"] == 0))

#call DF.iloc[passmask,"new_column_name] = what you want column equal to 
sub.loc[mask, 'match_diff_race'] = 1
#call DF.iloc[~passmask,"new_column_name] = what you want column equal to
#tilde to say not equal to condition, so that you can have the "opposite" classification
sub.loc[~mask, 'match_diff_race'] = 0


/Users/ltdanp21/anaconda/lib/python2.7/site-packages/pandas/core/indexing.py:461: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


In [6]:
columns_list = list(df.columns)
#print columns_list

In [7]:
sub.drop(['match',"samerace"], axis=1, inplace=True)

/Users/ltdanp21/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


In [8]:
sub.head()
# 1 = will date outside race
# 0 = will not date outside race
# risk = race may not have been an issue, simply did not vibe with the person?



gender  imprelig  goal  date  go_out  exphappy  expnum  attr1_1  sinc1_1  \
0       0       4.0   2.0   7.0     1.0       3.0     2.0     15.0     20.0   
1       0       4.0   2.0   7.0     1.0       3.0     2.0     15.0     20.0   
2       0       4.0   2.0   7.0     1.0       3.0     2.0     15.0     20.0   
3       0       4.0   2.0   7.0     1.0       3.0     2.0     15.0     20.0   
4       0       4.0   2.0   7.0     1.0       3.0     2.0     15.0     20.0   

   intel1_1  fun1_1  amb1_1  shar1_1 match_diff_race  
0      20.0    15.0    15.0     15.0               0  
1      20.0    15.0    15.0     15.0               0  
2      20.0    15.0    15.0     15.0               0  
3      20.0    15.0    15.0     15.0               1  
4      20.0    15.0    15.0     15.0               1

In [9]:
sub["match_diff_race"].value_counts()

0    7564
1     814
Name: match_diff_race, dtype: int64

In [10]:
sub.columns

Index([u'gender', u'imprelig', u'goal', u'date', u'go_out', u'exphappy',
       u'expnum', u'attr1_1', u'sinc1_1', u'intel1_1', u'fun1_1', u'amb1_1',
       u'shar1_1', u'match_diff_race'],
      dtype='object')

In [11]:
sub.isnull().sum()

gender                0
imprelig             79
goal                 79
date                 97
go_out               79
exphappy            101
expnum             6578
attr1_1              79
sinc1_1              79
intel1_1             79
fun1_1               89
amb1_1               99
shar1_1             121
match_diff_race       0
dtype: int64

gender                0

imprelig             79

goal                 79

date                 97

go_out               79

exphappy            101

expnum             6578

attr1_1              79

sinc1_1              79

intel1_1             79

fun1_1               89

amb1_1               99

shar1_1             121

match_diff_race       0

dtype: int64

In [12]:
sub.dropna(inplace = True)

/Users/ltdanp21/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


In [13]:
sub.isnull().sum()

gender             0
imprelig           0
goal               0
date               0
go_out             0
exphappy           0
expnum             0
attr1_1            0
sinc1_1            0
intel1_1           0
fun1_1             0
amb1_1             0
shar1_1            0
match_diff_race    0
dtype: int64

Now, conduct a PCA using scikit learn

http://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html

First, let's define the x and y variables: Airport is going to be our "x" variable

Then, standardize the x variable for analysis

In [14]:
#convert normalized columns back to a dataframe
x_norm = pd.DataFrame(x_std, columns = ['gender', 'imprelig', 'goal', 'date', 'go_out', 'exphappy',
       'expnum', 'attr1_1', 'sinc1_1', 'intel1_1', 'fun1_1', 'amb1_1',
       'shar1_1'])

NameError: name 'x_std' is not defined

In [ ]:
#convert response column to dataframe
y_df = pd.DataFrame(y, columns = ["match_diff_race"])

Next, create the covariance matrix from the standardized x-values and decompose these values to find the eigenvalues and eigenvectors

In [ ]:
sns.heatmap(x_norm.corr())

In [ ]:
x_norm.corr()

In [ ]:
# What will this function return and in what order? 
#Will return eigenValues first and eigenvectors second 
#pass df.corr to np.linalg.eig() 
eigenValues, eigenVectors = np.linalg.eig(x_norm.corr())

Then, check your eigenvalues and eigenvectors:

To find the principal componants, find the eigenpairs, and sort them from highest to lowest. 

In [ ]:
sorted_eigenVectors = np.sort(eigenVectors)   

In [ ]:
sorted_eigenValues = np.sort(eigenValues)

Next, Calculate the explained variance

In [ ]:
# figure this out...
eigen_total = sum(eigenValues)

def eigen_score(x):
   ExplainedVariance = (eigenValues[x]/eigen_total) * 100
   return "Explained Variance", x+1, ExplainedVariance

Finally, conduct the PCA - use the results about to guide your selection of "n" componants

In [15]:
x2 = sub[['gender', 'imprelig', 'goal', 'date', 'go_out', 'exphappy',
       'expnum', 'attr1_1', 'sinc1_1', 'intel1_1', 'fun1_1', 'amb1_1',
       'shar1_1']]

y2 = sub[["match_diff_race"]]

In [17]:
from sklearn.preprocessing import StandardScaler

#normalize predictor columns-instantiate function
ss = StandardScaler()

#pass predictor columns through normalizer
x_std2 = ss.fit_transform(x2)

In [18]:
from sklearn.decomposition import PCA

pca2 = PCA()

test_pca = pca2.fit_transform(x_std2)

In [19]:
pca2.explained_variance_ratio_

array([ 0.20102096,  0.13545705,  0.11169525,  0.08996495,  0.08787558,
        0.07525819,  0.06792594,  0.06085155,  0.05294239,  0.04871263,
        0.03607487,  0.02980453,  0.00241612])

In [20]:
values = pca2.components_

In [36]:

# just to check
evalue, evector = np.linalg.eig(pd.DataFrame(x_std2).corr().values)
a, b, c = np.linalg.svd(pd.DataFrame(x_std2).corr().values)

pairs = [[eva, eve] for eva, eve in zip(evalue, evector)]
pairs = sorted(pairs, reverse=True)

print [p[0] for p in pairs]
#print [p[0]/np.sum(evalue) for p in pairs]
#print [p[0] for p in pairs]
#print pca.explained_variance_

sk_comps = pd.DataFrame(pca2.components_)
eigvecs = pd.DataFrame(a.T)



[2.6132724530805427, 1.7609416692072219, 1.4520382592970791, 1.1695443032608455, 1.1423825418615861, 0.97835640879219943, 0.88303719648711376, 0.79107011709032293, 0.68825107698733545, 0.63326415938444025, 0.46897333164480681, 0.3874589379704016, 0.031409544936107793]


In [37]:
pca2.explained_variance_

array([ 2.61327245,  1.76094167,  1.45203826,  1.1695443 ,  1.14238254,
        0.97835641,  0.8830372 ,  0.79107012,  0.68825108,  0.63326416,
        0.46897333,  0.38745894,  0.03140954])

In [38]:
sk_comps

0         1         2         3         4         5         6   \
0   0.297442 -0.260138  0.016799 -0.349198 -0.221230  0.137536  0.314907   
1  -0.337432  0.192843 -0.453440 -0.326096 -0.231418  0.217361  0.397580   
2   0.261494 -0.231727  0.097862  0.218699  0.119154  0.402037  0.136330   
3  -0.395012 -0.213154  0.062549  0.067216  0.315351 -0.440236  0.163871   
4   0.040697  0.102624 -0.057920 -0.134780  0.199287 -0.121683  0.212114   
5  -0.208908 -0.357038  0.162725 -0.238490 -0.409424  0.130884  0.128820   
6  -0.101194 -0.019924  0.513674  0.030077 -0.573946 -0.338077 -0.104718   
7  -0.044005 -0.636675 -0.584756  0.092765 -0.125110 -0.230619 -0.307078   
8  -0.129737 -0.426281  0.259834  0.264352  0.166056  0.418221  0.106061   
9  -0.286892  0.259756 -0.191778  0.471772 -0.350971  0.340339 -0.254968   
10  0.317094  0.012511 -0.014988 -0.414539  0.007452  0.131432 -0.580669   
11 -0.557257 -0.048526  0.200738 -0.411729  0.283864  0.266391 -0.341327   
12  0.074539 -0.023408 -0.013123 -0.031737  0.007669  0.000180 -0.023184   

          7         8         9         10        11        12  
0   0.516838 -0.277575 -0.092861  0.027921 -0.330043 -0.303496  
1  -0.144592 -0.201000 -0.147902  0.205687  0.365438  0.138201  
2  -0.040748  0.390333 -0.626547  0.215615  0.177899  0.037171  
3   0.161657  0.001108 -0.154625  0.107514  0.319403 -0.553509  
4   0.269139  0.115847 -0.220657 -0.790317  0.181459  0.273320  
5  -0.253283  0.537355  0.281462 -0.263638  0.041397 -0.202060  
6   0.057909 -0.204672 -0.341967  0.048567  0.173741  0.274208  
7   0.037312 -0.027800 -0.118236 -0.030308 -0.076345  0.236932  
8  -0.003736 -0.487772  0.291240 -0.155001  0.279548  0.159363  
9   0.315778  0.002749 -0.112671 -0.240192 -0.021243 -0.341014  
10  0.051036 -0.071943  0.022882 -0.037368  0.555301 -0.233799  
11  0.110659  0.023159 -0.241080  0.042832 -0.337024  0.151408  
12 -0.654420 -0.376450 -0.371539 -0.341370 -0.229479 -0.338976

In [48]:
# The "components" in the PCA are essentially the mappings or weightings that will allow us to 
# transform our original coordinate system into the new principal component coordinate system.
#
# Every principal component is a linear combination of our original variables. Some variables
# from our original data may be more or less important to deriving the NEW coordinate on
# a given principal component axis.
#
# For example, going back to the dating example, PC1 (shallowness) is going to be a combination
# of the three questions we asked: values physical attractiveness, likes money, and 
# enjoys good food. Our "component" vector for PC1 is going to convert their answers on the
# 3 questions to get a value for shallowness, which is our new axis. Originally perhaps 
# the "x-axis" was "likes money". Now the x-axis is shallowness.
#
# Alright, so to get a person's shallowness value, we will multiply each component by each
# variable column. The vector of components for PC1: [c1, c2, c3] are weights for each column
# in our data. So:
# X_row1 = [attractiveness1, money1, food1]
# To get shallowness, we multiply the component vector across columns and add them together:
# shallowness1 = (c1 * attractiveness1) + (c2 * money1) + (c3 * food1)
#
# So in our new PCA matrix, which will have the same number of dimensions as our original,
# we can put that value in for the PC1 column:
#
# PC1             PC2    PC3
# [shallowness1, ...., .....]
# [    ....    , ...., .....]
#
# To get the PC2 and PC3, whatever they may represent, for this first person, we will need
# to multiply THEIR component vectors across that first row as well and summing them. Once
# we have the value we can fill it in for PC2 and PC3.
#
# Remember: all principal component variables are linear combinations of our original 
# variables, just in different ways. So shallowness may be 
# 0.4*money + 0.7*attactiveness + 0.01*food, but PC2 will have different weights for
# each of the columns since PC2 MUST BE UNRELATED TO PC1!
#
# Perhaps PC2 stands for "love of finding cheap food", for example. The equation for PC2
# might then be:
# PC2 = (0.7*money) + (0.01*attractiveness) + (0.5*food)
#
# Because liking money and liking food mean you like cheap but good food.
#
# Notice how the shallowness PC1 variable and the cheap food PC2 variable are created from
# the same columns in our original dataset, but in different ways. For shallowness the food
# column is not very important at all. For cheap food the attractiveness column is not
# very important at all. 
#
# Furthermore, being a shallow person and liking cheap food are character traits that are
# unrelated to each other. PCA guarantees that the new variables are not related.
#
# let's get the components out of the sklearn PCA object:
components = pca2.components


[ 0.29744168 -0.33743177  0.26149425 -0.39501211  0.04069665 -0.20890833
 -0.1011943  -0.04400474 -0.12973707 -0.28689216  0.31709398 -0.55725699
  0.07453901]


In [49]:
# These are the original coordinates for the first row/observation along your standardized variables.
# So, these are the attributes for "person 1" on the dimensions we originally measured her on. I'll print out
# the columns too for reference:
print x2.columns
print x_std2[0,:]

Index([u'gender', u'imprelig', u'goal', u'date', u'go_out', u'exphappy',
       u'expnum', u'attr1_1', u'sinc1_1', u'intel1_1', u'fun1_1', u'amb1_1',
       u'shar1_1'],
      dtype='object')
[-1.02043344  0.03553658 -0.16338008  1.34513361 -0.87415579 -1.53383692
 -0.74963845 -0.66687713  0.43409795 -0.24261819 -0.35912626  0.72236313
  0.6216566 ]


In [51]:
# This is your first person transformed into the principal component coordinate system that PCA finds.
# It's important to note here that the first value/column PC1 does NOT correspond directly to "gender"
# from above. It is a new variable comprised of some weighted combination of all our original variables.
# So, if this were shallowness, for example, it's some combination of gender, goal, go_out, etc. All the
# original variables contribute to this shallowness rating, to different degrees.
print test_pca[0,:]
#
print 'Just PC1:', test_pca[0,0]

[-1.91860049 -0.1214413  -0.39708761  0.53065742  0.15286599  0.24653608
  1.39188958  1.03065442 -0.35901289  0.62069226 -0.45271811 -0.59054281
  0.00220306]
Just PC1: -1.91860049167


In [56]:
# Kind of confusing because the variable is called "Principal Component 1", but the weighting vector used to 
# combine the original columns is also labeled as "component". Sorry. In class I called the
# component vector the eigenvector. It's eseentially the vector of weights that blueprints the transformation
# of your original columns into the singular PC1 value.
# 
# Let's print out the component vector for PC1 below:
print components[0, :]

# Note here that the components matrix that comes out of sklearn has the component vectors
# for each principal component organized in rows. So: rows are the vectors for each principal 
# component, and columns are the weights you multiply each of our original columns by before
# summing.

[ 0.29744168 -0.26013794  0.01679921 -0.34919753 -0.22123015  0.1375355
  0.31490658  0.51683777 -0.27757533 -0.09286115  0.02792137 -0.33004347
 -0.30349594]


In [58]:
# How do we get one value for our new variable PC1 from our old variables?
# Remember that the new variable (PC1) is some weighted combination of our
# old variables...
# 
# So we have the original variables for person 1, which I'll print again.
# We also have the component vector for PC1, which are the weights for our
# original columns – we will multiply each value for this person in each
# original column by it's corresponding "weight" in the component vector for
# PC1.
#
# Lastly, we sum together all these "reweighted" values of our original 
# variables into the NEW value for variable PC1.

print 'ORIGINAL COLUMN/VARIABLE VALUES:'
person1 = pd.DataFrame({'original_value':x_std2[0,:]}, index=x2.columns)
print person1
print '-----------------------------------\n'

print "MATCH WITH PC1'S COMPONENT VECTOR (WEIGHTS FOR ORIGINAL VARS):"
person1['PC1_component_vector'] = components[0, :]
print person1
print '-----------------------------------\n'

print "MULTIPLY ORIGINAL VAR VALUES WITH COMPONENT VECTOR WEIGHTS:"
person1['component_pieces_of_PC1'] = person1.original_value*person1.PC1_component_vector
print person1
print '-----------------------------------\n'

print "SUM THE PC1 COMPONENT PIECES TOGETHER TO GET VALUE FOR NEW VAR PC1:"
print 'person 1 PC1 value:',person1.component_pieces_of_PC1.sum()
print '-----------------------------------\n'

print 'So again, the new PC1 variable is some linear combination of our original variables!\n'

print 'We can get this same value out of the PCA object you fit on your original data:'
print test_pca[0,0]
print '...Which is the first row of the first column, aka PC1 value for person 1.'


np.sum(x_std2[0,:]*components.T[:,0])

ORIGINAL COLUMN/VARIABLE VALUES:
          original_value
gender         -1.020433
imprelig        0.035537
goal           -0.163380
date            1.345134
go_out         -0.874156
exphappy       -1.533837
expnum         -0.749638
attr1_1        -0.666877
sinc1_1         0.434098
intel1_1       -0.242618
fun1_1         -0.359126
amb1_1          0.722363
shar1_1         0.621657
-----------------------------------

MATCH WITH PC1'S COMPONENT VECTOR (WEIGHTS FOR ORIGINAL VARS):
          original_value  PC1_component_vector
gender         -1.020433              0.297442
imprelig        0.035537             -0.260138
goal           -0.163380              0.016799
date            1.345134             -0.349198
go_out         -0.874156             -0.221230
exphappy       -1.533837              0.137536
expnum         -0.749638              0.314907
attr1_1        -0.666877              0.516838
sinc1_1         0.434098             -0.277575
intel1_1       -0.242618             -0.092861


-1.9186004916669255

In [ ]:
# Now lets get the first row

In [ ]:
col = list(x2.columns)

In [ ]:
values_df = pd.DataFrame(values, columns = col)


In [ ]:
values_df

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.cross_validation import cross_val_score


logistic2 = LogisticRegression()

In [ ]:
y2b = np.ravel(y2)
y2b = y2b.astype(float)

In [ ]:
cvs = cross_val_score(logistic2, x_std2, y2b, cv = 5 )

In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt
% matplotlib inline